File: py/75.颜色分类.py
#
# @lc app=leetcode.cn id=75 lang=python3
#
# [75] 颜色分类
#

# @lc code=start
class Solution:
    def sortColors(self, nums) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        p0, cur, p2 = 0, 0, len(nums)-1
        
        while cur <= p2:
            
            if nums[cur] == 0:
                nums[p0], nums[cur] = nums[cur], nums[p0]
                p0 += 1
                cur += 1
            elif nums[cur] == 2:
                nums[p2], nums[cur] = nums[cur], nums[p2]
                p2 -= 1
            else:
                cur += 1
            
        return nums


# @lc code=end

i = [2,0,1]
print(Solution().sortColors(i))

File: golang/75.颜色分类.go
package main

import "fmt"

/*
 * @lc app=leetcode.cn id=75 lang=golang
 *
 * [75] 颜色分类
 */

// @lc code=start
func sortColors(nums []int) []int {
	if len(nums) == 0 {
		return nums
	}
	p0, cur, p2 := 0, 0, len(nums)-1
	for cur <= p2 {
		switch nums[cur] {
		case 0:
			nums[p0], nums[cur] = nums[cur], nums[p0]
			p0++
			cur++
		case 2:
			nums[p2], nums[cur] = nums[cur], nums[p2]
			p2--
		default:
			cur++
		}
	}
	return nums
}

// @lc code=end

func main() {
	i := []int{2, 0, 1}
	fmt.Println(sortColors(i))
}
